app/backend/approaches/chatreadretrieveread.py

```diff
@@ -282,14 +282,18 @@ def get_messages_from_history(
 
         message_builder.append_message(self.USER, user_content, index=append_index)
 
-        for h in reversed(history[:-1]):
-            if message_builder.token_length > max_tokens:
-                break
-            if bot_msg := h.get("bot"):
-                message_builder.append_message(self.ASSISTANT, bot_msg, index=append_index)
-            if user_msg := h.get("user"):
-                message_builder.append_message(self.USER, user_msg, index=append_index)
-
+        newest_to_oldest = list(reversed(history[:-1]))
+        for ind in range(len(newest_to_oldest)):
+            message = newest_to_oldest[ind]
+            if message.get("role") == self.ASSISTANT:
+                ind += 1
+                while ind < len(newest_to_oldest) and newest_to_oldest[ind].get("role") != self.USER:
+                    ind += 1
+                prev_message = newest_to_oldest[ind]
+                if message_builder.calculate_token_length_with_messages([message, prev_message]) > max_tokens:
+                    break
+                message_builder.append_message(self.ASSISTANT, message["content"], index=append_index)
+                message_builder.append_message(self.USER, prev_message["content"], index=append_index)
         return message_builder.messages
```

Comment:
@pamelafox thanks for taking care of this. I had same solution in mind. But we can implement  in a different way since this iterates over 2 loops and for 32K limit, conversation length might be longer and cause larger runtime.

My thinking is that
1. Track token count of messages in chatapproach.
2. As we currently do, start adding latest message to the message builder. 
3. Before we add a message into message builder, evaluate token length of the new message and evaluate the  condition
4. `msg_count = messagebuilder.cal_token_length_for_msg(new message)`
    ` if msg_count + total_token_count > max_tokens: break`
    `msg_builder.append(new message)`
     `total_token_count  = messagebuilder.get_tokencount()`
    This would avoid two loops and still follow LIFO to add the message builder.

Reply:
I've updated the code to look more like your suggestion. However, I still have it so that it only adds a *pair* at a time, so it always checks for bot/user pairs. That seems like how it was originally and I thought it was intentional to always return pairs or none at all. If we think its helpful to return a bot's answer even without the user's question, then I can adjust it. (That would be simpler code!)

Reply:
Thats fine I think, model needs the latest context. ref: https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/chatgpt?pivots=programming-language-chat-completions#managing-conversations

Reply:
@pamelafox also we need to reduce maxtokens to maxtokens - 1024. Because the token limit error says Model limit reached with (3K in messages and 1024 in completion) So we need to account for the max response token length returned from model which we hardcoded in the code aas 1024.

Reply:
I've changed it so it's fine with breaking up message pairs, which is much simpler code. Also added another test for that. And I made the change to consider max token length as subtracting 1024.

Reply:
Also added a test for the 1024 issue, plus a logging.debug statement for when that's reached, as that seems like it might be helpful for folks debugging their message histories.